### PR TITLE
Implement HPUHooksInterface

### LINKED ISSUE #137262: [regression] PT 2.5 does not support memory pinning for out-of-tree accelerators as Intel Gaudi
### 🐛 Describe the bug

The introduction of new AcceleratorHookInterface (PR #126376) broke pinning memory for out-of-tree Intel Gaudi accelerator. The implementation took care of in-tree devices and PrivateUse1.
As a result, the following error is seen on Intel Gaudi:
`RuntimeError: HPU device type not an accelerator`

We are working to prepare a missing patch to fix Intel Gaudi support.

### Versions

Collecting environment information...
PyTorch version: 2.5.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.30.4
Libc version: glibc-2.35

Python version: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.1.85+-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 535.104.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               2
On-line CPU(s) list:                  0,1
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) CPU @ 2.20GHz
CPU family:                           6
Model:                                79
Thread(s) per core:                   2
Core(s) per socket:                   1
Socket(s):                            1
Stepping:                             0
BogoMIPS:                             4399.99
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            32 KiB (1 instance)
L1i cache:                            32 KiB (1 instance)
L2 cache:                             256 KiB (1 instance)
L3 cache:                             55 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0,1
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Vulnerable; SMT Host state unknown
Vulnerability Meltdown:               Vulnerable
Vulnerability Mmio stale data:        Vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Vulnerable
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:             Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Vulnerable (Syscall hardening enabled)
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] optree==0.12.1
[pip3] torch==2.5.0+cpu
[pip3] torchaudio==2.4.1+cu121
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.19.1+cu121
[conda] Could not collect

cc @bdhirsh @gujinghui @EikanWang @fengyuan14 @guangyey

### PR DESCRIPTION
**Pull request by tbohutyn**
Fixes #137262

### CODE REVIEW

**Comment by bsochack** [comment]
@albanD please help to review. The change fixes a critical regression after interface changes

File: aten/src/ATen/detail/HPUHooksInterface.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#include <ATen/detail/HPUHooksInterface.h>
+#include <c10/util/CallOnce.h>
+
+namespace at {
+namespace detail {
```
Comment (malfet) [inline aten/src/ATen/detail/HPUHooksInterface.cpp:5]: Please use nested namespace
```suggestion
namespace at::detail {
```

**Review by albanD — COMMENTED** [review]
Sorry about that!
The long term plan is that we will add full regression testing for the privateuse1 backend and we can prevent this kind of issues for these out of tree users.

File: aten/src/ATen/DeviceAccelerator.cpp [hunk 2]
```diff
@@ -43,6 +43,7 @@ bool isAccelerator(c10::DeviceType d) {
     case at::kXPU:
     case at::kHIP:
     case at::kMPS:
+    case at::kHPU:
```
[the remark this answers is not in the document]
Comment (guangyey) [reply aten/src/ATen/DeviceAccelerator.cpp:46]: Yes, please refer to https://github.com/pytorch/pytorch/blob/395e098209ddbb5780fe96eddc853c05c0709d17/aten/src/ATen/DeviceAccelerator.cpp#L29

File: aten/src/ATen/Context.h [hunk 3]
```diff
@@ -41,6 +42,7 @@ class TORCH_API Context {
     c10::DeviceType device_type = device.type();
     initCUDAIfNeeded(device_type);
     initHIPIfNeeded(device_type);
+    initHPUIfNeeded(device_type);
```
Comment (guangyey) [inline aten/src/ATen/Context.h:45]: Just a reminder: here will conflict with https://github.com/pytorch/pytorch/pull/136519, please rebase this PR if https://github.com/pytorch/pytorch/pull/136519 is merged first.

File: aten/src/ATen/detail/HPUHooksInterface.cpp (hunk 1 above)
Comment (tbohutyn) [reply aten/src/ATen/detail/HPUHooksInterface.cpp:5]: not everything is in detail, so I'll leave it as is

**Review by guangyey — APPROVED** [review]
Reminder: would you like to update the [doc](https://github.com/pytorch/pytorch/blob/f9ed39c9891f3b487b7816509e6be59e907890c2/docs/source/torch.rst?plain=1#L155) as well?

File: aten/src/ATen/Context.h [hunk 4]
```diff
@@ -462,6 +467,10 @@ inline bool hasXPU() {
   return globalContext().hasXPU();
 }
 
+inline bool hasHPU() {
+  return globalContext().hasHPU();
```
Comment (guangyey) [inline aten/src/ATen/Context.h:471]: globalContext needs to add the `hasHPU` method.

**Comment by tbohutyn** [comment]
@guangyey @albanD can I get a review and merge on updated PR?

**Comment by guangyey** [comment]
> @guangyey @albanD can I get a review and merge on updated PR?

Please fix the code conflicts.

**Comment by tbohutyn** [comment]
Refactor of Context.h was reverted, seems like it will be applied back soon:
https://github.com/pytorch/pytorch/pull/136519
Waiting until it is merged to avoid resolving conflicts back and forth

**Comment by malfet** [comment]
Removing 2.5.1 milestone as this feels like a feature work

File: aten/src/ATen/detail/HPUHooksInterface.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    return false;
+  }
+
+  Allocator* getPinnedMemoryAllocator() const override {
+    TORCH_CHECK(
+        false,
+        "You should register `HPUHooksInterface` for HPU before call `getPinnedMemoryAllocator`.");
+  }
+
+  bool hasPrimaryContext(C10_UNUSED DeviceIndex device_index) const override {
```
Comment (guangyey) [inline aten/src/ATen/detail/HPUHooksInterface.h:43]:
```suggestion
  bool hasPrimaryContext([[maybe_unused]] DeviceIndex device_index) const override {
```
`C10_UNUSED` is deprecated.

File: aten/src/ATen/detail/HPUHooksInterface.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  void init() const override {
+    TORCH_CHECK(false, "Cannot initialize HPU without HPU backend");
+  }
+
+  virtual bool hasHPU() const {
+    return false;
+  }
+
+  const Generator& getDefaultGenerator(
+      C10_UNUSED DeviceIndex device_index = -1) const override {
```
Comment (guangyey) [inline aten/src/ATen/detail/HPUHooksInterface.h:24]:
```suggestion
      [[maybe_unused]] DeviceIndex device_index = -1) const override {
```

**Review by guangyey — APPROVED** [review]
Some small nits. Otherwise LGTM.
You still need a final stamp from the core maintainers.

**Comment by kit1980** [comment]
2.5.1 is an emergency patch release to address specific large regressions, moving this to 2.6.0

**Review by albanD — APPROVED** [review]
Sounds ok.
Note that this makes HPU an accelerator device so this has quite a big implication and you're going to get a lot of automatic features you need to handle from it (pin memory, device-generic api, etc).

**Comment by bsochack** [comment]
> Sounds ok. Note that this makes HPU an accelerator device so this has quite a big implication and you're going to get a lot of automatic features you need to handle from it (pin memory, device-generic api, etc).

Yes. This is also where we noticed regression on PT 2.5 as it became mandatory to implement.

**Comment by guangyey** [comment]
@tbohutyn please resolve the conflicts with `aten/src/ATen/Context.h`

**Pull request closed without merging**